Make the "Save dive data as subtitles" feature more configurable.

vigge93 opened the pull request:
Previously, the subtitle generation was hardcoded, making it unsuitable if you didn't want all of the displayed values. This has been replaced by a format string that is configurable in the settings, using predefined tags that are replaced with the values. The default value for this has been set to (mostly) match the currently generated subtitle string. This also provides a good starting point for users that want to modify the string.

<!-- Lines like this one are comments and will not be shown in the final output. -->
<!-- Make sure that you have read the "Contributing" section of the README and also the notes in CodingStyle. -->
<!-- If you are a collaborator, please add labels and assign other collaborators for a review. -->

### Describe the pull request:
<!-- Replace [ ] with [x] to select options. -->
- [ ] Bug fix
- [X] Functional change
- [X] New feature
- [ ] Code cleanup
- [ ] Build system change
- [ ] Documentation change
- [ ] Language translation

### Pull request long description:
Previously, the subtitle generation was hardcoded, making it unsuitable if you didn't want all of the displayed values. This has been replaced by a format string configurable in the settings, using predefined tags replaced with the corresponding values. The default value for this has been set to (mostly) match the currently generated subtitle string. This also provides a good starting point for users who want to modify the string. Further, a small fix has been added where the subtitles now start at the beginning of the video, and not at the first datapoint after the video starts. A table of the available tags has also been added directly below the input field.

### Changes made:
<!-- Enumerate the changes with 1), 2), 3) etc. -->
<!-- Ensure the test cases are updated if needed. -->
1) Added "Subtitle format string" and "Available tags" under a new "Dive data as subtitles" section in the Media preferences.
2) Added `subtitles_format_string` to `preferences` struct.
3) Updated `format_st_event` to use the `subtitles_format_string` and replace the tags.
4) Added translation messages to translation source.

### Release note:
<!-- Describe if this change needs a release note present in CHANGELOG.md. -->
<!-- Also, please make sure to add the release note on top of the file CHANGELOG.md. -->
Release note has been added at top of `CHANGELOG.md`

### Documentation change:
<!-- If this PR makes changes to user functionality, then the documentation has to be updated too. -->
<!-- Please, briefly outline here what has changed in terms of the user experience (UX). -->
<!-- If UX changes have been made, a maintainer should apply the 'needs-documentation-change' label. -->
New setting in media preferences to specify the subtitle format string.

mikeller commented on translations/subsurface_source.ts:
```diff
@@ -7258,6 +7258,186 @@ See https://doc.qt.io/archives/qt-4.8/qdatetime.html#fromString</source>
         <source>One can specify te relative position within the video from where the thumbnail is retrieved. Use the slider below do do that.</source>
         <translation type="unfinished"></translation>
     </message>
+    <message>
```
The translations are extracted automatically, so this part is not needed (and might interfere with adding the generated translations).

vigge93 replied:
Alright, I have reverted the changes.

atdotde commented:
From what I see this is a great improvement (if in fact there are people that use this feature)! Thanks a lot.

mikeller commented on core/save-profiledata.cpp:
```diff
@@ -175,32 +187,214 @@ static std::string format_st_event(const plot_data &entry, const plot_data &next
 	const char *unit;
-
-	if (entry.sec < offset || entry.sec > offset + length)
+	if (next_entry.sec < offset || entry.sec > offset + length)
```
Why are we relaxing this condition here? Presumably if `entry.sec` is smaller than `offset` we are outside of the video timeline, even if `next_entry.sec` is not smaller than `offset`.

vigge93 replied:
It's the fencepost problem. If we have one datapoint at 35:20 and one at 35:40, but the video starts at 35:30, we need to use the data from the 35:20 datapoint for the first 10 seconds of the clip.

mikeller commented on core/save-profiledata.cpp:
```diff
@@ -175,32 +187,214 @@ static std::string format_st_event(const plot_data &entry, const plot_data &next
+	}
+
+	if (entry.icd_warning) {
+		replace_all(format_string, "[icd_warning]", translate("gettextFromC", "ICD in leading tissue"));
```
Seems to be a bit wordy, I am worried that the subtitles will be displayed in a very small font if this is ever hit during a dive.

vigge93 replied:
I agree that it may be a bit verbose. Something like "ICD warning" might be better. However, "ICD in leading tissue" is used in the plot, so I don't know if we want to keep the same formulation for consistency?

mikeller reviewed (approved):
I like the idea.

However, it looks like there's still a bug of some sort in this, a in a quick test the placeholders did not get updated for me:

![Screenshot 2024-10-24 161906](https://github.com/user-attachments/assets/675c3a5e-df58-47e6-8f41-bb1ef618a548)

Testing the same dive / video with current master was working fine:

![Screenshot 2024-10-24 162413](https://github.com/user-attachments/assets/60bf7744-5d06-4553-bfd5-9e391baa10bf)

I also think that it would make sense to add a mention of this to the user manual (or even move the full list of supported placeholders there): https://github.com/subsurface/subsurface/blob/1a9b09682ad4bb99a4526734c5b8041f884d47af/Documentation/user-manual.txt#L1817

mikeller reviewed (changes requested):
See comments.

vigge93 commented:
> I like the idea.
> 
> However, it looks like there's still a bug of some sort in this, a in a quick test the placeholders did not get updated for me:
> 
> ![Screenshot 2024-10-24 161906](https://private-user-images.githubusercontent.com/4742747/379536506-675c3a5e-df58-47e6-8f41-bb1ef618a548.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjk3NDg3NTQsIm5iZiI6MTcyOTc0ODQ1NCwicGF0aCI6Ii80NzQyNzQ3LzM3OTUzNjUwNi02NzVjM2E1ZS1kZjU4LTQ3ZTYtOGY0MS1iYjFlZjYxOGE1NDgucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTAyNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEwMjRUMDU0MDU0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZjU4ZWRlZGQ0MDk3OTZiOWY2MzhmMWM3YTQ5OWI5OGYwM2RjNmFjZjVlY2VmYTE5N2IxZmVjYjA0MThiMThjYiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.WSv6x9eRWdFG0rJQgdHhgNpZxFRwsFDh8mJuI00fupU)
> 
> Testing the same dive / video with current master was working fine:
> 
> ![Screenshot 2024-10-24 162413](https://private-user-images.githubusercontent.com/4742747/379536573-60bf7744-5d06-4553-bfd5-9e391baa10bf.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjk3NDg3NTQsIm5iZiI6MTcyOTc0ODQ1NCwicGF0aCI6Ii80NzQyNzQ3LzM3OTUzNjU3My02MGJmNzc0NC01ZDA2LTQ1NTMtYmZkNS05ZTM5MWJhYTEwYmYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTAyNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEwMjRUMDU0MDU0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZDQ3YTE2OTQ1ZTBlYmVmZTI0NTIwZjc0ZGZkOWYxYWFiMWEyOGYyOGY1MTIzNTkxNGYwM2NlZjU1YjJjOGJjYiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.5VwcbCKmfMuPivBHNLhHttwt5UlDx-FUAvzNsbk74zc)
> 
> I also think that it would make sense to add a mention of this to the user manual (or even move the full list of supported placeholders there):
> 
> https://github.com/subsurface/subsurface/blob/1a9b09682ad4bb99a4526734c5b8041f884d47af/Documentation/user-manual.txt#L1817

Yes, a small bug managed to sneak its way into the `replace_all` function while I was cleaning up the code. It has been fixed now. I have also updated the user manual and added the list of tags. I still think there is value in keeping it in the settings as well as a quick reference. Is there any way to make it collapsible so it doesn't have to be present all of the time?

bstoeger commented on core/save-profiledata.cpp:
```diff
@@ -45,6 +47,16 @@ static std::string video_time(int secs)
 	return format_string_std("%d:%02d:%02d.000,", hours, mins, secs);
 }
 
+void replace_all(std::string& str, const std::string& old_value, const std::string& new_value) {
```
I know it's a "religious" issue, but as per coding-style, we put the`&` with the variable, not the type. Rationale: That's how the C++ grammar works, like it or not.

Also I suggest making this a global helper function.

If not a global helper, it should be of static linkage.

vigge93 replied:
Fixed. Made the function static.

bstoeger commented on core/save-profiledata.cpp:
```diff
@@ -45,6 +47,16 @@ static std::string video_time(int secs)
 }
 
+void replace_all(std::string& str, const std::string& old_value, const std::string& new_value) {
+    if (old_value.empty())
```
This check seems unnecessary to me. If `old_value` is empty, the loop body will simply not execute. The common case will be a non-empty string, so no point in wasting code on this.

vigge93 replied:
If `old_value` is empty, it will always match at `start_pos`, resulting in it getting stuck in an infinite loop.

bstoeger replied:
Yes correct - I thought this was testing the modified string.

mikeller commented:
@vigge93:

> I still think there is value in keeping it in the settings as well as a quick reference. Is there any way to make it collapsible so it doesn't have to be present all of the time?

Makes sense. For now there is plenty of space available in the 'Media preferences' tab. If we ever add more stuff there we can always remove or shorten the list, and put in a reference to the manual.
I am not a Qt expert, so I don't know how to do collapsible content in Qt.

mikeller reviewed (approved):
Thanks for fixing, this is working well now.

One thought I had when testing: It would be an improvement if it was possible to have an optional prefix for individual datapoints, at least for the ones that are not always displayed, so that something like 'ceiling=[ceiling]' would be hidden in its entirety if there was no ceiling. Maybe encoded as `{ceiling=[ceiling]}`. But this can be a future extension.

mikeller reviewed (approved):
Thanks for fixing, this is working well now.

One thought I had when testing: It would be an improvement if it was possible to have an optional prefix for individual datapoints, at least for the ones that are not always displayed, so that something like 'ceiling=[ceiling]' would be hidden in its entirety if there was no ceiling. Maybe encoded as `{ceiling=[ceiling]}`. But this can be a future extension.

vigge93 commented:
I think this should be ready to merge once the workflows have been run @mikeller.
